### PR TITLE
Spring 8

### PR DESCRIPTION
**Pull request by Fedich09**

### CODE REVIEW

**Review by MykolaFatkullin — APPROVED** [review]

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -43,7 +43,8 @@ public User add(User user) {
     @Override
     public Optional<User> findByEmail(String email) {
         try (Session session = sessionFactory.openSession()) {
-            Query<User> query = session.createQuery("SELECT u FROM User u WHERE u.email = :email",
+            Query<User> query = session.createQuery("SELECT u FROM User u "
+                            + "left join fetch u.roles WHERE u.email = :email",
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: use inner join here, all users should have roles

File: src/main/java/com/dev/cinema/dbinit/InjectData.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private void addRolesAndAdmin() {
+        Role adminRole = new Role();
+        adminRole.setRoleName("ADMIN");
+        Role userRole = new Role();
+        roleService.add(adminRole);
+        userRole.setRoleName("USER");
+        roleService.add(userRole);
+        User user = new User();
+        user.setEmail("admin@gmail.com");
+        user.setPassword(securityConfig.getEncoder().encode("admin"));
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dbinit/InjectData.java]: password should be encoded in the UserService

File: src/main/java/com/dev/cinema/model/User.java [hunk 3]
```diff
@@ -17,6 +20,16 @@
     @Column(unique = true)
     private String email;
     private String password;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/User.java]: it's lazy by default

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 4]
```diff
@@ -1,25 +1,36 @@
...
     }
 
     protected void configure(HttpSecurity httpSecurity) throws Exception {
-        httpSecurity.authorizeRequests()
+        httpSecurity
+                .authorizeRequests()
+                .antMatchers(HttpMethod.POST).hasRole("ADMIN")
+                .antMatchers(HttpMethod.PUT).hasRole("ADMIN")
+                .antMatchers(HttpMethod.DELETE).hasRole("ADMIN")
+                .antMatchers(HttpMethod.GET).permitAll()
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java:33]: is user allowed to perform complete order ?(it's POST)

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -10,16 +10,16 @@
 @Service
 public class UserServiceImpl implements UserService {
     private final UserDao userDao;
-    private final SecurityConfig securityConfig;
+    private final AppConfig appConfig;
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: inject Encoder here instead of Appconfig (it's a bean as well)

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 6]
```diff
@@ -31,6 +31,7 @@ protected void configure(HttpSecurity httpSecurity) throws Exception {
                 .antMatchers(HttpMethod.PUT).hasRole("ADMIN")
                 .antMatchers(HttpMethod.DELETE).hasRole("ADMIN")
                 .antMatchers(HttpMethod.GET).permitAll()
+                .antMatchers(HttpMethod.POST, "/orders/complete").hasRole("USER")
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java:34]: adding product to shopping cart is not GET as well

**Review by RomanDubovskyi — APPROVED** [review]
improve requests permissions according to Sofia's message in general chat

**Pull request merged**